File: api/proxy.py
from http.server import BaseHTTPRequestHandler
from urllib.parse import parse_qs, urlparse
from urllib.request import Request, urlopen
import json
import xml.etree.ElementTree as ET

MUFG_BASE = 'https://in.mpms.mufg.com/Initial_Offer/IPO.aspx'

HEADERS = {
    'Content-Type': 'application/json; charset=UTF-8',
    'Accept': 'application/json, text/plain, */*',
    'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36',
    'X-Requested-With': 'XMLHttpRequest',
    'Referer': 'https://in.mpms.mufg.com/Initial_Offer/IPO.aspx',
}


def get_mufg_list():
    """Scrapes the live IPO list from MUFG Registrar."""
    url = MUFG_BASE + '/GetDetails'
    req = Request(url, data=b'{}', headers=HEADERS, method='POST')
    with urlopen(req, timeout=25) as response:
        raw_xml = json.loads(response.read().decode('utf-8')).get('d', '')

    root = ET.fromstring(raw_xml)
    ipo_list = []
    for table in root.findall('Table'):
        company_id = table.find('company_id')
        company_name = table.find('companyname')
        if company_id is not None and company_name is not None and company_id.text and company_name.text:
            ipo_list.append({
                'id': company_id.text,
                'name': company_name.text,
            })
    return ipo_list


def mufg_search(data):
    endpoint = MUFG_BASE + '/SearchOnPan'
    payload = {
        'clientid': str(data.get('clientid', '')).strip(),
        'PAN': str(data.get('PAN', '')).strip().upper(),
        'IFSC': str(data.get('IFSC', '')),
        'CHKVAL': str(data.get('CHKVAL', '1')),
        'token': str(data.get('token', '')),
    }
    req = Request(endpoint, data=json.dumps(payload).encode('utf-8'), headers=HEADERS, method='POST')
    with urlopen(req, timeout=25) as response:
        return response.read()


class handler(BaseHTTPRequestHandler):
    def _cors(self):
        self.send_header('Access-Control-Allow-Origin', '*')
        self.send_header('Access-Control-Allow-Methods', 'GET, POST, OPTIONS')
        self.send_header('Access-Control-Allow-Headers', 'Content-Type, X-Requested-With')

    def _json(self, code, obj):
        body = json.dumps(obj).encode('utf-8')
        self.send_response(code)
        self.send_header('Content-Type', 'application/json; charset=utf-8')
        self._cors()
        self.send_header('Content-Length', str(len(body)))
        self.end_headers()
        self.wfile.write(body)

    def do_OPTIONS(self):
        self.send_response(204)
        self._cors()
        self.end_headers()

    def do_GET(self):
        parsed = urlparse(self.path)
        action = (parse_qs(parsed.query).get('action', [''])[0] or '').lower()
        if action == 'ipos':
            try:
                self._json(200, get_mufg_list())
            except Exception as e:
                self._json(502, {'error': f'MUFG Error: {e}'})
            return
        self._json(404, {'error': 'Not found'})

    def do_POST(self):
        content_length = int(self.headers.get('Content-Length', 0))
        body = self.rfile.read(content_length) if content_length else b'{}'

        try:
            data = json.loads(body.decode('utf-8') or '{}')
        except Exception:
            data = {}

        action = str(data.get('action', '')).lower()
        if action == 'ipos':
            try:
                self._json(200, get_mufg_list())
            except Exception as e:
                self._json(502, {'error': f'MUFG Error: {e}'})
            return

        try:
            result = mufg_search(data)
            self.send_response(200)
            self.send_header('Content-Type', 'application/json')
            self._cors()
            self.end_headers()
            self.wfile.write(result)
        except Exception as e:
            self._json(502, {'error': str(e)})

File: api/proxy.js
const MUFG_BASE = 'https://in.mpms.mufg.com/Initial_Offer/IPO.aspx';

const HEADERS = {
  'Content-Type': 'application/json; charset=UTF-8',
  Accept: 'application/json, text/plain, */*',
  'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36',
  'X-Requested-With': 'XMLHttpRequest',
  Referer: 'https://in.mpms.mufg.com/Initial_Offer/IPO.aspx',
};

function cors(res) {
  res.setHeader('Access-Control-Allow-Origin', '*');
  res.setHeader('Access-Control-Allow-Methods', 'GET, POST, OPTIONS');
  res.setHeader('Access-Control-Allow-Headers', 'Content-Type, X-Requested-With');
}

function parseIpos(rawXml) {
  const ipos = [];
  const tableRegex = /<Table>([\s\S]*?)<\/Table>/g;
  let match;
  while ((match = tableRegex.exec(rawXml)) !== null) {
    const block = match[1];
    const id = block.match(/<company_id>(.*?)<\/company_id>/)?.[1];
    const name = block.match(/<companyname>(.*?)<\/companyname>/)?.[1];
    if (id && name) ipos.push({ id, name });
  }
  return ipos;
}

async function getMufgList() {
  const response = await fetch(`${MUFG_BASE}/GetDetails`, {
    method: 'POST',
    headers: HEADERS,
    body: '{}',
  });
  const data = await response.json();
  return parseIpos(data.d || '');
}

async function mufgSearch(body) {
  const payload = {
    clientid: String(body.clientid || '').trim(),
    PAN: String(body.PAN || '').trim().toUpperCase(),
    IFSC: String(body.IFSC || ''),
    CHKVAL: String(body.CHKVAL || '1'),
    token: String(body.token || ''),
  };
  return fetch(`${MUFG_BASE}/SearchOnPan`, {
    method: 'POST',
    headers: HEADERS,
    body: JSON.stringify(payload),
  });
}

module.exports = async (req, res) => {
  cors(res);
  if (req.method === 'OPTIONS') {
    res.status(204).end();
    return;
  }

  const action = (req.query && req.query.action) || (req.body && req.body.action) || '';

  if (req.method === 'GET' && action === 'ipos') {
    try {
      res.status(200).json(await getMufgList());
    } catch (error) {
      res.status(502).json({ error: `MUFG Error: ${error.message || error}` });
    }
    return;
  }

  if (req.method === 'POST') {
    if (String(action).toLowerCase() === 'ipos') {
      try {
        res.status(200).json(await getMufgList());
      } catch (error) {
        res.status(502).json({ error: `MUFG Error: ${error.message || error}` });
      }
      return;
    }

    try {
      const upstream = await mufgSearch(req.body || {});
      const text = await upstream.text();
      res.status(upstream.status);
      res.setHeader('Content-Type', upstream.headers.get('content-type') || 'application/json');
      res.end(text);
    } catch (error) {
      res.status(502).json({ error: String(error.message || error) });
    }
    return;
  }

  res.status(404).json({ error: 'Not found' });
};
